### ISSUE TITLE
left shift of signed UB?

### ISSUE THREAD

**Issue by mattkretz** [mattkretz/wg21-papers#29]
@jensmaurer:
> if E1 has a signed type and non-negative value, and E1×2E2 is representable in the corresponding unsigned type of the result type, then that value, converted to the result type, is the resulting value; otherwise, the behavior is undefined. [5.8 p2]

Since P0214 inherits the operator semantics from the `value_type` ("_Returns:_ A `datapar` object initialized with the results of the component-wise application of the indicated operator."), left shift of signed values that would shift into or over the sign bit *are* UB.
Do we want to make it implementation-defined in P0214? Making it well-defined would require two's complement, right?
Otherwise I'll make my implementation assume that left-shift of signed types can never overflow. Invoking UB would thus shift bits in the neighboring element.

IMHO: I'm fine with inheriting UB from the `value_type`. It's a QoI issue to diagnose incorrect usage. Do we want guidance from LEWG on this?

**Comment by jensmaurer** [comment]
My opinion: If you feel you must perform shifts with "signed int", do it, but live with the consequences (i.e. undefined behavior if touching the sign bit). I understand people are often using shifts for a "fast multiply by power-of-2", but I consider this misguided, because the compiler will optimize a multiply to a shift anyway, and if you want a multiplication by n, you should actually write "x*n" in your source code and not some obfuscating shift. That means I'd rather prohibit shifts on "signed int" entirely, but I'm 40+ years late asking for that.

In general, we have quite a few new types in the pipeline (e.g. extended integer typers), and LEWG + EWG + SG6 should form an opinion whether they want to retain the cruft in certain areas that maybe was once useful to be compatible with long-outdated hardware, or whether they want to come up with a new set of rules in that area. Maybe even mandating two's complement for datapar.

In short: Don't special-case the shift operators for datapar (except maybe for the scalar overload issue).

**Comment by mattkretz** [comment]
https://godbolt.org/g/qEgCZ2 OK, since the compiler can do the shift instead of mul optimization here I'm sympathetic to your position. The main reason for having a signed shift would be to generate an arithmetic right shift instead of a logic right shift. However, C++ makes negative right shift implementation-defined, so I'm not sure how value this is either.

You think LEWG wants to have a say? Or should I just remove signed shifts from the paper? After removal we can more easily add functionality later without breaking API/ABI.

**Comment by jensmaurer** [comment]
Just remove signed shifts from the paper, and wait for LEWG to want them desperately.

**Comment by mattkretz** [comment]
> wait for LEWG to want them desperately

here we are. LEWG feedback: Wants operator<<(signed) to work except where it's undefined for the underlying integer.

@jensmaurer IIUC then datapar is to inherit the UB 1:1 from the builtin type, right?
